akka-http-core/src/main/scala/akka/http/impl/util/StreamUtils.scala

```diff
@@ -260,16 +260,20 @@ private[http] object StreamUtils {
    * INTERNAL API
    */
   @InternalApi
-  object CaptureMaterializationAndTerminationOp extends EntityStreamOp[(Future[Unit], Future[Unit])] {
-    def strictM: (Future[Unit], Future[Unit]) = (Future.successful(()), Future.successful(()))
-    def apply[T, Mat](source: Source[T, Mat]): (Source[T, Mat], (Future[Unit], Future[Unit])) = {
+  object CaptureMaterializationAndTerminationOp extends EntityStreamOp[(Future[Unit], Future[Unit], Future[Option[UniqueKillSwitch]])] {
+    def strictM: (Future[Unit], Future[Unit], Future[Option[UniqueKillSwitch]]) = (Future.successful(()), Future.successful(()), Future.successful(None))
+    def apply[T, Mat](source: Source[T, Mat]): (Source[T, Mat], (Future[Unit], Future[Unit], Future[Option[UniqueKillSwitch]])) = {
       val materializationPromise = Promise[Unit]()
+      val killSwitchPromise = Promise[Option[UniqueKillSwitch]]()
+      val newSource0 =
+        source.via(Flow.fromGraph(KillSwitches.single[T]).mapMaterializedValue(ks => killSwitchPromise.trySuccess(Some(ks))))
```

Comment:
Do we expect adding the killswitch here to affect performance?

Reply:
Probably not more than the already termination capturing, but maybe I could fold it into the existing stage. In any case, cost is amortized to some degree at the OS receive buffers. We always read up to the configured direct buffer size from the OS so if you read less you'll get more data in the next run.

Reply:
I'll do this improvement in a follow-up PR.